pkg/api/service.go

```diff
@@ -76,31 +106,75 @@ func (s Service) loadChart(chartName string) (*chart.Chart, error) {
 	return requestedChart, nil
 }
 
-func (s Service) installChart(icfg InstallConfig, ch *chart.Chart, vals chartValues) (*InstallResult, error) {
-	s.installer.SetConfig(icfg)
-	release, err := s.installer.Run(ch, vals)
+func (s Service) installChart(icfg ReleaseConfig, ch *chart.Chart, vals ChartValues) (*ReleaseResult, error) {
+	s.Installer.SetConfig(icfg)
+	release, err := s.Installer.Run(ch, vals)
 	if err != nil {
 		return nil, fmt.Errorf("error in installing chart: %v", err)
 	}
-	result := new(InstallResult)
+	result := new(ReleaseResult)
 	if release.Info != nil {
-		result.status = release.Info.Status.String()
+		result.Status = release.Info.Status.String()
 	}
 	return result, nil
 }
 
-func (s Service) validate(icfg InstallConfig, values chartValues) error {
+func (s Service) upgradeRelease(ucfg ReleaseConfig, ch *chart.Chart, vals ChartValues) (*ReleaseResult, error) {
+	s.upgrader.SetConfig(ucfg)
+	release, err := s.upgrader.Run(ucfg.Name, ch, vals)
+	if err != nil {
+		return nil, fmt.Errorf("error in upgrading chart: %v", err)
+	}
+	result := new(ReleaseResult)
+	if release.Info != nil {
+		result.Status = release.Info.Status.String()
+	}
+	return result, nil
+}
+
+func (s Service) validate(icfg ReleaseConfig, values ChartValues) error {
 	if strings.HasPrefix(icfg.ChartName, ".") ||
 		strings.HasPrefix(icfg.ChartName, "/") {
 		return errors.New("cannot refer local chart")
 	}
 	return nil
 }
 
-func NewService(settings *cli.EnvSettings, cl chartloader, i installer) Service {
-	return Service{
-		settings:    settings,
-		chartloader: cl,
-		installer:   i,
+func (s Service) List(releaseStatus string) ([]Release, error) {
+	listStates := new(action.ListStates)
+
+	state := action.ListAll
+	if releaseStatus != "" {
+		state = listStates.FromName(releaseStatus)
+	}
+
+	if state == action.ListUnknown {
+		return nil, errors.New("invalid release status")
 	}
+
+	s.lister.SetState(state)
+	s.lister.SetStateMask()
+	releases, err := s.lister.Run()
+	if err != nil {
+		return nil, err
+	}
+
+	var helmReleases []Release
+	for _, eachRes := range releases {
```

Comment:
just call it `rel`, or even `r`, it's fine in go than readable variables for loop